### PR TITLE
kernel: Introduce k_triggered_work

### LINKED ISSUE #10748: Work waiting on pollable objects
At the moment Zephyr has two kinds of works - executed straightaway or executed with delay. It would be useful to have a third kind in which work would be executed when specific pollable object is ready (or any from a given list).

This way one could avoid creating a new thread to wait for some blocking API to become ready. Since creation of a thread costs RAM it may be to expensive to create many threads for every small task that use blocking read or something similar.

This could be solved by creating a dedicated thread that waits for any of the registered objects and when ready submit an associated work.
Or this could be hooked right into the scheduler - when object is ready and there is any work associated with it, such work would be submitted for execution.

### PR DESCRIPTION
**Pull request by pizi-nordic**
This PR introduces triggered work. The triggered work is submitted to workqueue when one of watched events happen. Such mechanism allows to eliminate costly dedicated threads to execute something when data is available or some resource become ready.

The work is driven mostly by device power management (details: https://github.com/zephyrproject-rtos/zephyr/pull/13382#issuecomment-467910416) and @pdunaj needs.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/10748

### CODE REVIEW

**Review by andyross — COMMENTED** [review]
Maybe it would be a little more general to have a callback-based API where a kernel function can be registered on a pollable object that gets called on event state changes, then implement k_triggered_work() in terms of that?  That strikes me as useful too, and that way we won't have to implement it as two separate subsystems.

**Comment by pabigot** [comment]
@bfrog and @microbuilder might be able to provide feedback as to whether this feature would help solve sensor notification problems where https://github.com/zephyrproject-rtos/zephyr/issues/13718#issuecomment-467131339 suggested that callbacks are problematic but we still want to get rid of coupling sensor drivers to a specific thread hand-off solution.

I don't fully understand @andyross's suggestion but if its intent is to support immediate execution of specific actions ("kernel function") on event notifications, where one of the available actions is submitting work to a queue, I could see that being more general.  Examples of other useful triggers would help motivate the generalization.

**Comment by pdunaj** [comment]
@andyross 

>  a callback-based API where a kernel function can be registered on a pollable object that gets called on event state changes

I am not sure this would work well. If state change happens from atomic context (e.g. interrupt) some API (e.g. mutex) would be unusable. Triggering work with pollable objects makes code executed from workq thread which has fewer limitations. Triggered work should work in userspace.

As an alternative I was also thinking about creating callback manager that would have its own thread and poll on behalf of others. This is however more expressive from the RAM perspective as you have to create yet another thread and yet another stack.

**Comment by pizi-nordic** [comment]
I have updated the PR introducing a callback into `_poller` structure. I think that this is the thing suggested by @andyross as it allows for easy extension in the future.

I have kept the callback internal to the polling code, as exporting it to the user might create a lot of trouble: It is called with interupts disabled from various contexts.

**Comment by andyross** [comment]
To be clear: my thinking here is targetted more at how to evolve poll instead of bolting things onto it.  IMHO poll (not really the subject of this patch per se) is implemented at the wrong level of abstraction.  Threads that wait for something, in Zephyr, are always waiting on a wait_q.  So the wait_q **should** be 1:1 with our "signalable event" thing.  But that's not the way poll works, so everything needs to be special cased ("Oh, you're waiting on a semaphore, that's different than being pended on a fifo and I'm going to handle them with a big switch statement").

I'd like to get those switches and special cases out of poll, and make "signal interest in multiple wait_q's" a single lower level operation.  In that world, having special cases on what to do **after** the signal (i.e. ready the thread vs. add a work item) has all the same problems and I'd like to have the low level operation be as decoupled from the API as possible.

**Comment by pizi-nordic** [comment]
@andyross: What you propose will not work with the k_triggered_work interface introduced here. I simply do not have a thread which I could put in the wait_q.

File: kernel/poll.c [hunk 1]
```diff
@@ -188,36 +188,78 @@ static inline void set_event_ready(struct k_poll_event *event, u32_t state)
...
+	}
+
+	_ready_thread(thread);
+
+	return 0;
+}
+
+int _impl_k_poll(struct k_poll_event *events, int num_events, s32_t timeout)
+{
+	__ASSERT(!_is_in_isr(), "");
```
Comment (tbursztyka) [inline kernel/poll.c]: asserts after the variable declarations

File: kernel/poll.c [hunk 2]
```diff
@@ -448,3 +475,170 @@ Z_SYSCALL_HANDLER1_SIMPLE_VOID(k_poll_signal_reset, K_OBJ_POLL_SIGNAL,
...
+	work->poller.thread = NULL;
+	work->real_handler = handler;
+}
+
+int k_triggered_work_submit_to_queue(struct k_work_q *work_q,
+				     struct k_triggered_work *work,
+				     struct k_poll_event *events,
+				     int num_events)
+{
+	__ASSERT(work_q != NULL, "NULL work_q\n");
```
Comment (tbursztyka) [inline kernel/poll.c]: same

File: kernel/poll.c (hunk 1 above)
Comment (pizi-nordic) [reply kernel/poll.c]: Fixed.

File: kernel/poll.c (hunk 2 above)
Comment (pizi-nordic) [reply kernel/poll.c]: Fixed

**Comment by pizi-nordic** [comment]
@andyross, @andrewboie: Any comments to the implementation?

**Comment by andyross** [comment]
Getting back from vacation and leaving a drive-by note: this looks good to me design-wise but I still need to do a real review.  Please nag me post-1.14-release.

**Comment by carlescufi** [comment]
@andyross nagging you post 1.14 release :)

**Comment by andrewboie** [comment]
> Triggered work should work in userspace.

None of these new APIs appear to have user mode support.
What's the plan here?

**Comment by pizi-nordic** [comment]
>>Triggered work should work in userspace.

> None of these new APIs appear to have user mode support.
What's the plan here?

At the moment there are no plans to add userspace support here. Also, as far as I know whole k_*work* API is not accessible from userspace, so adding such support will require fixing that part too.

**Comment by andrewboie** [comment]
> Also, as far as I know whole k_work API is not accessible from userspace, so adding such support will require fixing that part too

Workqueues can run in user mode, see k_work_q_user_start() and k_work_submit_to_user_queue(). The design is not optimal and some refactoring is going on for 1.15, but there is support there.

**Comment by pizi-nordic** [comment]
I have looked into the code and I do not see a fast & east way to add usermode support, so probably I will seek you on slack in order to ask some questions. So far this one is most important for me: Is there a way to distinguish user mode and kernel mode work queue (or attached thread)?

**Comment by pizi-nordic** [comment]
@andrewboie: I spend some time looking how we can add userspace support to this. Unfortunately I have not found optimal solution.

The problem is that `k_work` structure has be accessible from the user-space workq and somehow linked with other members `k_triggered_work` structure. So:

- If we make `k_triggered_work` a userspace object, the kernel will rely on information from userspace which is a serious vulnerability.
- If we make `k_triggered_work` a kernel object, the userspace workq will trigger exception while accessing embedded `k_work` structure.

As result the only way solve this puzzle is to have `k_triggered_work_descriptor` as a kernel object keeping all the internal stuff with link into userspace `k_work`. However such approach will not match the current `k_work/k_delayed_work` API look & feel.

Summarising: I would like to keep this PR as is and have this API in kernel only at the moment (similarly to existing `k_delayed_work`). Later, if we would like to have full API in userspace, we can drive the `k_delayed_work` and `k_triggered_work` towards descriptor-based approach.

**Comment by andrewboie** [comment]
Thanks for taking the time for looking into it.

File: include/kernel.h [hunk 3]
```diff
@@ -3040,6 +3057,112 @@ static inline s32_t k_delayed_work_remaining_get(struct k_delayed_work *work)
...
+ *
+ * @param work_q Address of workqueue.
+ * @param work Address of delayed work item.
+ * @param events An array of pointers to events which trigger the work.
+ * @param num_events The number of events in the array.
+ *
+ * @retval 0 Work item started watching for events.
+ * @retval -EINVAL Work item is being processed or has completed its work.
+ * @retval -EADDRINUSE Work item is pending on a different workqueue.
+ * @req K-TWORK-001
```
Comment (pabigot) [inline include/kernel.h]: remove `@req` annotations throughout.

File: include/kernel.h [hunk 4]
```diff
@@ -3040,6 +3057,112 @@ static inline s32_t k_delayed_work_remaining_get(struct k_delayed_work *work)
...
+ * Submitting a previously submitted triggered work item that is still
+ * waiting for the event cancels the existing submission and reschedules it
+ * the using the new event list. Note that this behavior is inherently subject
+ * to race conditions with the pre-existig triggered work item and work queue,
+ * so care must be taken to synchronize such resubmissions externally.
+ *
+ * @note Can be called by ISRs.
+ *
+ * @warning
+ * Provided array of events as well as a triggered work item must not be
```
Comment (pabigot) [inline include/kernel.h]: This should also make clear that, unlike `k_poll`, the event array and the events it references must be in global (possibly heap-allocated) memory, not thread/stack memory.

File: kernel/poll.c [hunk 5]
```diff
@@ -188,36 +188,77 @@ static inline void set_event_ready(struct k_poll_event *event, u32_t state)
...
 				__ASSERT(false, "unexpected return code\n");
 			}
 		}
 		k_spin_unlock(&lock, key);
 	}
 
+	return events_registered;
+}
+
+static int k_pool_poller_cb(struct k_poll_event *event, u32_t state)
```
Comment (pabigot) [inline kernel/poll.c]: `k_pool` is confusing; was this meant to be named `k_poll_poller_cb`?

File: kernel/poll.c [hunk 6]
```diff
@@ -172,11 +172,11 @@ static inline void clear_event_registration(struct k_poll_event *event)
 
 /* must be called with interrupts locked */
 static inline void clear_event_registrations(struct k_poll_event *events,
-					      int last_registered,
-					      k_spinlock_key_t key)
+					     int num_events,
+					     k_spinlock_key_t key)
 {
```
Comment (pabigot) [inline kernel/poll.c]: None of these changes have any significance, so they really should be in a separate commit or left out of the PR.

File: kernel/poll.c [hunk 7]
```diff
@@ -188,36 +188,77 @@ static inline void set_event_ready(struct k_poll_event *event, u32_t state)
...
-
-	/* find events whose condition is already fulfilled */
-	for (int ii = 0; ii < num_events; ii++) {
+	for (int i = 0; i < num_events; i++) {
+		k_spinlock_key_t key;
 		u32_t state;
 
 		key = k_spin_lock(&lock);
-		if (is_condition_met(&events[ii], &state)) {
-			set_event_ready(&events[ii], state);
```
Comment (pabigot) [inline kernel/poll.c]: Again lots of green and red all due to changes in variable names, obscuring the important differences.

**Review by pabigot — DISMISSED** [review]
I needed something like this for my simplified sensor infrastructure, where an external thread animates a state machine that needs to transition on events such as signals.  In my case I do that with the main thread and `k_poll` for a signal that's triggered by a timeout or interrupt callback.  It works well.

I'm not convinced this abstraction is going to get us very far for things like power management and device initialization, though.  If the next step that's released by the event includes a bus transaction, like SPI, the calling thread (now a work queue thread) will still block.

So I'm ambivalent about this without seeing a bit more detail in #10748 describing exactly what use cases this capability satisfies and how to use it.

If it does seem worth doing, some bikeshedding from a Zephyr architecture consistency perspective:

Looking at this I believe `k_delayed_work...` was misnamed, and should have been `k_work_delayed...` so that all things related to the concept of work queue have a common prefix.

It's too late to fix that, but to avoid further fragmentation the new API including functions and data structures should start with a prefix that reflects the infrastructure it re-uses.  So I propose that this should be `k_work_poll...` or `k_poll_work...` rather than `k_triggered_work...`.

Opposing or supporting views, anybody?

For the main review:

There seem to be two primary capability changes in this PR:

* `k_poll` is reworked to separate managing a set of poll events from processing fact-of-event, where the reaction is now provided by a kernel callback held in the poller object.
* A new API is added that uses a set of poll events to queue work rather than wake a thread.

These two should be separated into two commits to simplify review and future work (i.e. make it more clear what has to be done to add a new "do-this-on-event" API).

There are also changes to control structures (`for` => `while`) and variable names (`ii` => `i`, `last_registered` => `events_registered` or `num_events`) that appear to have no impact on behavior.  It would be nice (I'm not insisting) if these were reverted if they aren't necessary, or if they have clear benefit moved to a separate cleanup commit that has no functional effect, so reviewers can focus on what actually changed.

The `@req` annotations should be removed; I see some came in already with `k_delayed_work` (could be removed in a cleanup commit). Requirements are awesome, but wherever the referenced ones are tracked it's not in Zephyr, so they add no value.

Finally the kernel documentation needs to be updated to describe this functionality and show how it would be used.  The only existing example is deep in the test code, and is of no use to somebody who wants to make use of the feature.  Which gets back to "why, specifically, do we need this?".

File: include/kernel.h (hunk 3 above)
Comment (pizi-nordic) [reply include/kernel.h]: Done.

File: kernel/poll.c (hunk 5 above)
Comment (pizi-nordic) [reply kernel/poll.c]: Yes. Thank you for spotting that.

File: kernel/poll.c (hunk 6 above)
Comment (pizi-nordic) [reply kernel/poll.c]: Fixed.

File: kernel/poll.c (hunk 7 above)
Comment (pizi-nordic) [reply kernel/poll.c]: Fixed

File: include/kernel.h (hunk 4 above)
Comment (pizi-nordic) [reply include/kernel.h]: Done.

**Comment by pizi-nordic** [comment]
> I'm not convinced this abstraction is going to get us very far for things like power management and device initialization, though. If the next step that's released by the event includes a bus transaction, like SPI, the calling thread (now a work queue thread) will still block.
>
> So I'm ambivalent about this without seeing a bit more detail in #10748 describing exactly what use cases this capability satisfies and how to use it.

It is true, that the current and foreseen problems addressed by this change might be not fully visible for someone reading this PR, so I will try to put some sight on motivation behind this work.

Currently the design of the Zephyr frameworks implies existence of the multiple small threads because API exposed by most of the peripherals/subsystems is blocking. As each thread needs stack, this scheme requires a lot of RAM for the thread stacks.

The situation is slowly changing as more and more APIs are designed to support asynchronous model of operation by exposing to user a poll-able object, which allows single thread to service multiple entities (the SPI is an example of asynchronous device API).

This PR goes step further: It eliminates need of "dedicated" threads handling devices, significantly reducing the RAM usage. In the proposed model, the poll-able object state changes triggers execution of associated work.

I personally see this useful in device idle power management: The API waking up the device could schedule wake-up for all the dependencies first, then, when all notifications signalling ready status arrive, wake-up itself sending ready status to the application at the end. In this case the workqueue will execute only the FSM, which should not require too much resources and should lead to any sleeping.

File: include/kernel.h (hunk 3 above)
Comment (pabigot) [reply include/kernel.h]: I'm still seeing them, but it looks like there are a lot of pre-existing ones, so leaving the new ones in too doesn't seem worth fussing about.

File: include/kernel.h (hunk 4 above)
Comment (pabigot) [reply include/kernel.h]: I don't see this either.  Did something get dropped?

File: include/kernel.h (hunk 3 above)
Comment (pizi-nordic) [reply include/kernel.h]: Grrrr. I not staged kernel.h to commit. Now it should be OK.

File: include/kernel.h (hunk 4 above)
Comment (pizi-nordic) [reply include/kernel.h]: Yes, I missed kernel.h when I prepared new commit.  Now it should be OK.

**Comment by pizi-nordic** [comment]
>>I personally see this useful in device idle power management: The API waking up the device could schedule wake-up for all the dependencies first, then, when all notifications signalling ready status arrive, wake-up itself notifying the application at the end. In this case the workqueue will execute only the FSM and functions initiating device state transition. This could be done also for devices on SPI bus as the driver itself could use spi_transceive_async() to schedule transfer changing device power state.

>This is true, but for it to work in that role we need a clearly stated expectation that k_work_handler_t functions (at least ones used in this way) will never block or busy-wait.

I thought that this is clear for any user of `k_work` and friends as we should never block system work queue.

>To do that we need much better/consistent handling of asynchronous operations in Zephyr. For example, I2C has no async API, and spi_transceive_async() will block if there's already an asynchronous operation in progress (#15703). Solving this generally becomes complex because for each driver class we need a structure that captures the parameters of the requested operations, and for each driver instance a queue of pending operations, possibly with some sort control over potential interleaving of complex transactions from different clients.

So far we have asynchronous device idle power management API, which cannot be implemented for some kinds of devices for the reasons you mentioned. IMHO we eventually need something like
[TWI transaction manager](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk5.v15.3.0%2Fgroup__nrf__twi__mngr.html&cp=5_1_6_11_55) and [SPI transaction manager](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk5.v15.3.0%2Fgroup__nrf__spi__mngr.html&cp=5_1_6_11_46) for any kind of the shared busses as well as better drivers interface, as any blocking in the fact a waste of resources in memory constrained devices.

>In any case, I think my main concern about justifying this PR would be addressed by clear documentation of how to use the new feature to solve a simple problem. The rest is a different problem.

I will add small chapter describing k_triggered_work usage to the docs.

**Comment by pizi-nordic** [comment]
>> In any case, I think my main concern about justifying this PR would be addressed by clear documentation of how to use the new feature to solve a simple problem. The rest is a different problem.

>I will add small chapter describing k_triggered_work usage to the docs.

Done.

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: doc/reference/kernel/threads/index.rst [hunk 8]
```diff
@@ -557,6 +557,40 @@ queue, unless the work item has already been removed and processed by the
 workqueue's thread. Consequently, once a work item's timeout has expired
 the work item is always processed by the workqueue and cannot be canceled.
 
+Triggered Work
+==============
+
+The triggered work schedules a work item in response on a **poll event**
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply doc/reference/kernel/threads/index.rst]: Updated.

File: doc/reference/kernel/threads/index.rst [hunk 9]
```diff
@@ -557,6 +557,40 @@ queue, unless the work item has already been removed and processed by the
...
+(see: :ref:polling_v2), allowing to execute user defined function when
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply doc/reference/kernel/threads/index.rst]: Updated.

File: doc/reference/kernel/threads/index.rst [hunk 10]
```diff
@@ -557,6 +557,40 @@ queue, unless the work item has already been removed and processed by the
...
+observed resource become available or a poll signal is raised.
+In contrast to :cpp:func:`k_poll()`, the triggered work does not require
+dedicated thread waiting or actively polling for a **poll event**.
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply doc/reference/kernel/threads/index.rst]: Udated

File: doc/reference/kernel/threads/index.rst [hunk 11]
```diff
@@ -557,6 +557,40 @@ queue, unless the work item has already been removed and processed by the
...
+
+A triggered work item is a standard work item that has the following
+added properties:
+
+* A pointer to array with poll events specifying list of events which
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply doc/reference/kernel/threads/index.rst]: Updated.

File: doc/reference/kernel/threads/index.rst [hunk 12]
```diff
@@ -557,6 +557,40 @@ queue, unless the work item has already been removed and processed by the
...
+  will trigger work item submission to the workqueue's queue.
+
+* A size of the array containing poll events.
+
+A triggered work item is initialized and submitted to a workqeueue in a similar
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply doc/reference/kernel/threads/index.rst]: Fixed.

File: doc/reference/kernel/threads/index.rst [hunk 13]
```diff
@@ -557,6 +557,40 @@ queue, unless the work item has already been removed and processed by the
...
+manner to a standard work item, although dedicated kernel APIs are used.
+When submit request is made, the kernel begins to observe kernel objects
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply doc/reference/kernel/threads/index.rst]: Updated

File: doc/reference/kernel/threads/index.rst [hunk 14]
```diff
@@ -557,6 +557,40 @@ queue, unless the work item has already been removed and processed by the
...
+specified by the poll events. Once at least one of the observed kernel
+objects change state, the work item is submitted to the specified workqueue,
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply doc/reference/kernel/threads/index.rst]: Updated.

File: doc/reference/kernel/threads/index.rst [hunk 15]
```diff
@@ -557,6 +557,40 @@ queue, unless the work item has already been removed and processed by the
...
+where it remains pending until it is processed in the standard manner.
+
+.. important::
+    The triggered work item as well as referenced array of poll events
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply doc/reference/kernel/threads/index.rst]: Updated.

**Comment by pizi-nordic** [comment]
@dbkinder, @pabigot: I have updated the PR some time ago. Please re-review.

**Review by pabigot — COMMENTED** [review]
My earlier concerns about motivation have been adequately addressed within the scope of this PR.  I've now used this (actually its predecessor `k_delayed_work()`) in my sensple API and believe it will help, though we also desperately need a generic event notification infrastructure that has a pollable variant.

My remaining concern is this, which I won't use to block the PR.  But I still want to see opinions from others before approving.

> Looking at this I believe `k_delayed_work...` was misnamed, and should have been `k_work_delayed...` so that all things related to the concept of work queue have a common prefix.
> 
> It's too late to fix that, but to avoid further fragmentation the new API including functions and data structures should start with a prefix that reflects the infrastructure it re-uses. So I propose that this should be `k_work_poll...` or `k_poll_work...` rather than `k_triggered_work...`.
> 
> Opposing or supporting views, anybody?

**Comment by galak** [comment]
Dev-review:
* Looking for additional review
* Rename api to be `k_work_` prefix
* @pabigot to email dev list for additional review.

**Comment by pabigot** [comment]
In a last run through of this I finally identified part of why this approach hasn't seemed quite right, and it relates to my original question about cases where this would be used.

`k_poll` supports waiting for a state change event occurs with an optional limit on wait time, and it blocks the calling thread until either event occurs.

`k_delayed_work()` supports waiting until a time limit is reached, then schedules a work item.  No thread blocking, no support for state change triggers.

This pr introduces `k_work_triggered` which supports waiting until a state change event occurs, then schedules a work item.  No thread blocking, no support for wait time limits.

This isn't solving the whole problem, because I still have no way to wait until either an event *or* an alarm occurs without blocking a thread.

Why don't we make this:
```
int k_work_poll(struct k_work_poll* work,
                struct k_poll_event *events,
		size_t num_events,
		k_timeout_t timeout); /* s32_t timeout for now */
```

which always returns immediately but for non-error situations will store the return code that `k_poll` would have returned (indicating either timeout or event) into a field in `struct k_work_poll`, so the work item can handle both timeout and event triggers.

This seems most general, and allows us to replace `k_delayed_work` with a call to `k_work_poll` with `num_events` set to zero.

Thoughts?

**Review by dbkinder — APPROVED** [review]
+1 for doc changes, thanks.
Sorry for the re-review delay.

**Comment by pizi-nordic** [comment]
> Why don't we make this:
>```
>int k_work_poll(struct k_work_poll* work,
> struct k_poll_event *events,
>		size_t num_events,
>		k_timeout_t timeout); /* s32_t timeout for now */
>```

I like the idea. I will try to add this to this PR.

**Comment by pizi-nordic** [comment]
@pabigot: I added the timeout, but so far this is proof of concept - there is no tests written yet. I pushed the changes for early review. I think that I will add the tests tomorrow (and fixes :)).

**Comment by pizi-nordic** [comment]
@pabigot: I added basic tests.

**Review by pabigot — COMMENTED** [review]
This seems ok.  There's enough that I'm not entirely happy about that I don't want to go ahead and approve it.  There's nothing here that I would, at this moment, insist be changed, but I don't think there's been enough eyes on it from a functional perspective to just merge it.

I'd still like to see it named `k_work_poll` since it's essentially a threadless `k_poll`.

Also the timeout addition should be squashed into the commit that adds the new API.

When a work item timeout occurs I believe it would be better to delay disabling the poller until the wrapper work handler clears the event registry.  That would require separate tracking of whether the timeout fired and whether events were discovered; e.g. add to `struct k_work_triggered` a `signaled` flag that gets set in `k_work_triggered_poller_cb` just as `expired` is set in `triggered_work_expiration_handler`.  Then the consumer can optimally handle any events that occur between the timeout and the point where the work handler gets invoked rather than assuming there's nothing to do except timeout activity.

@andyross or somebody from the kernel team should determine whether the addition of `CONFIG_POLL=y` to `tests/kernel/workq/work_queue` is appropriate, since that eliminates the former use of that test when `CONFIG_POLL` is not selected.

File: tests/kernel/workq/work_queue/prj.conf [hunk 16]
```diff
@@ -1,3 +1,4 @@
 CONFIG_ZTEST=y
 CONFIG_QEMU_TICKLESS_WORKAROUND=y
+CONFIG_POLL=y
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply tests/kernel/workq/work_queue/prj.conf]: Done

File: kernel/poll.c [hunk 17]
```diff
@@ -456,3 +482,208 @@ static inline void z_vrfy_k_poll_signal_reset(struct k_poll_signal *signal)
...
+
+	k_work_submit_to_queue(work_q, &twork->work);
+}
+
+static int k_work_triggered_poller_cb(struct k_poll_event *event, u32_t status)
+{
+	struct _poller *poller = event->poller;
+
+	if (poller->is_polling && poller->thread) {
+		struct k_work_triggered *work =
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply kernel/poll.c]: Done

File: kernel/poll.c [hunk 18]
```diff
@@ -456,3 +482,208 @@ static inline void z_vrfy_k_poll_signal_reset(struct k_poll_signal *signal)
...
+	struct k_work_q *work_q =
+		CONTAINER_OF(twork->poller.thread, struct k_work_q, thread);
+
+	twork->poller.is_polling = false;
+	twork->expired = true;
+
+	k_work_submit_to_queue(work_q, &twork->work);
+}
+
+static int k_work_triggered_poller_cb(struct k_poll_event *event, u32_t status)
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply kernel/poll.c]: Done

**Comment by pizi-nordic** [comment]
> When a work item timeout occurs I believe it would be better to delay disabling the poller until the wrapper work handler clears the event registry. That would require separate tracking of whether the timeout fired and whether events were discovered; e.g. add to struct k_work_triggered a signaled flag that gets set in k_work_triggered_poller_cb just as expired is set in triggered_work_expiration_handler. Then the consumer can optimally handle any events that occur between the timeout and the point where the work handler gets invoked rather than assuming there's nothing to do except timeout activity.

I think that it is possible now:
- If expired flag is false, the event happened before timeout.
- If expired flag true, the timeout happened before any event. However the events are still monitored after timeout up to clean-up phase. As result the application could scan the events to check if something changed state after timeout and before execution.

**Comment by pabigot** [comment]
> As result the application could scan the events to check if something changed state after timeout and before execution.

Yes, but it's more accurate to say that the application *must* scan the events, as once the timeout occurs nothing outside the event array records the fact of a subsequent signal.  In `k_poll` there's a binary choice between `0` and `-EAGAIN`, which is fine there because there is no significant interval between the timeout and clearing the event registration.  For the work item there's an arbitrary delay between those two steps which increases the likelihood of an event being signalled but not announced.

I had proposed:
> for non-error situations [`k_work_poll`] will store the return code that `k_poll` would have returned (indicating either timeout or event) into a field in `struct k_work_poll`, so the work item can handle both timeout and event triggers.

Switching from `bool expired` to `int poll_result` would seem to make sense as it's compatible with existing API.  Unfortunately that API doesn't support representing that both timeout and an event signal have occurred, though we could do a post-timeout change of the result from `-EAGAIN` to `0` when an event comes in, which would be pretty natural.  We could also handle `-EINTR` too.

Again I'm not requesting that you do this; I still want other eyes on it before approving.  As is it's a nice feature (though with surprisingly little interest); it could just have better compatibility with existing API, IMO.

**Comment by pabigot** [comment]
requested by @pizi-nordic for API meeting: provide input on https://github.com/zephyrproject-rtos/zephyr/pull/14041#issuecomment-535953527 i.e. whether API should follow `k_poll` in name and return value, or remain distinct as `k_work_triggered`.

Prototype in last commits of https://github.com/pabigot/zephyr/commits/pr/14041

attn @carlescufi

**Comment by pabigot** [comment]
Added to dev-review since https://github.com/zephyrproject-rtos/zephyr/pull/14041#issuecomment-537046410 didn't get time in API telecon.  @pizi-nordic would really like to get the API settled so this can be merged.

File: doc/reference/kernel/threads/workqueue.rst [hunk 19]
```diff
@@ -122,6 +122,41 @@ queue, unless the work item has already been removed and processed by the
 workqueue's thread. Consequently, once a work item's timeout has expired
 the work item is always processed by the workqueue and cannot be canceled.
 
+Triggered Work
+**************
+
+The :cpp:func:`k_work_poll()` interface schedules a triggered work
+item in response to a **poll event** (see :ref:`polling_v2`), that will
+call a user-defined function when the monitored resource becomes available
+or a poll signal is raised. In contrast to :cpp:func:`k_poll()`,
```
Comment (pabigot) [inline doc/reference/kernel/threads/workqueue.rst]: " ~the~ _a_ monitored resource becomes available or ~a~ poll signal is raised, _or a timeout occurs_."

File: doc/reference/kernel/threads/workqueue.rst [hunk 20]
```diff
@@ -122,6 +122,41 @@ queue, unless the work item has already been removed and processed by the
 workqueue's thread. Consequently, once a work item's timeout has expired
 the work item is always processed by the workqueue and cannot be canceled.
 
+Triggered Work
+**************
+
+The :cpp:func:`k_work_poll()` interface schedules a triggered work
```
Comment (pabigot) [inline doc/reference/kernel/threads/workqueue.rst]: This needs to be `k_work_poll_submit()` for the documentation link to work.

**Review by pabigot — CHANGES_REQUESTED** [review]
Two small documentation changes (fix link, mention submit on timeout).

File: doc/reference/kernel/threads/workqueue.rst (hunk 19 above)
Comment (pizi-nordic) [reply doc/reference/kernel/threads/workqueue.rst]: Fixed

File: kernel/poll.c [hunk 21]
```diff
@@ -188,36 +188,77 @@ static inline void set_event_ready(struct k_poll_event *event, u32_t state)
...
+
+int z_impl_k_poll(struct k_poll_event *events, int num_events, s32_t timeout)
+{
+	int events_registered;
+	k_spinlock_key_t key;
+	struct _poller poller = { .is_polling = true,
+				  .thread     = _current,
+				  .cb         = k_poll_poller_cb };
+
+	__ASSERT(!z_is_in_isr(), "");
```
Comment (pabigot) [inline kernel/poll.c]: Sorry, didn't run it first: this needs to be `z_arch_is_in_isr()` (upstream rename).

File: doc/reference/kernel/threads/workqueue.rst (hunk 20 above)
Comment (pizi-nordic) [reply doc/reference/kernel/threads/workqueue.rst]: Fixed

File: kernel/poll.c (hunk 21 above)
Comment (pizi-nordic) [reply kernel/poll.c]: Fixed.

**Review by pabigot — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**